src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+	while (bytes_already_read != file_size) {
+		const auto chunk_size = clamp_to_uint16(file_size - bytes_already_read);
+		uint16_t bytes_read = chunk_size;
+		if (!DOS_ReadFile(handle, content.data() + bytes_already_read, &bytes_read) ||
+		    bytes_read != chunk_size) {
+			DOS_CloseFile(handle);
+			return KeyboardLayoutResult::CpiReadError;
+		}
+		bytes_already_read += chunk_size;
+	}
+
+	DOS_CloseFile(handle);
+
+	const auto name_for_log = format_str("the '%s' file", file_name.c_str());
+	CpiParser parser(content, name_for_log);
+	const auto parser_result = parser.GetFonts(code_page);
+
+	if (parser_result.unsupported_cpx_file) {
+		return KeyboardLayoutResult::UnsupportedCpxFile;
+	} else if (!parser_result.status_ok) {
+		return KeyboardLayoutResult::InvalidCpiFile;
+	}
+
+	// Return the font
+	const auto& extracted_ega_fonts = parser_result.extracted_ega_fonts;
+	if (extracted_ega_fonts.contains(code_page)) {
+		const auto& font = extracted_ega_fonts.at(code_page);
+		check_screen_font_complete(font, code_page, name_for_log);
+		return font;
+	}
+	const auto& extracted_lcd_fonts = parser_result.extracted_lcd_fonts;
+	if (extracted_lcd_fonts.contains(code_page)) {
+		const auto& font = extracted_lcd_fonts.at(code_page);
+		check_screen_font_complete(font, code_page, name_for_log);
+		return font;
+	}
+	const auto& extracted_other_fonts = parser_result.extracted_other_fonts;
+	if (extracted_other_fonts.contains(code_page)) {
+		const auto& font = extracted_other_fonts.at(code_page);
+		check_screen_font_complete(font, code_page, name_for_log);
+		return font;
+	}
+
+	// Let the user know if screen font was found but could not be used
+	if (parser_result.found_ega_fonts.contains(code_page) ||
+	    parser_result.found_lcd_fonts.contains(code_page) ||
+	    parser_result.found_other_fonts.contains(code_page)) {
+		return KeyboardLayoutResult::ScreenFontUnusable;
+	}
+
+	// Let the user know if he used a printer CPI file
+	if (parser_result.found_printer_fonts && !parser_result.found_screen_fonts) {
+		return KeyboardLayoutResult::PrinterCpiFile;
+	}
+
+	// In all the other case - tell the user the screen font was not found
+	// TODO: List the screen code pages in the KEYB command output
+	auto all_found = parser_result.found_other_fonts;
+	std::merge(parser_result.found_ega_fonts.begin(),
+	           parser_result.found_ega_fonts.end(),
+	           parser_result.found_lcd_fonts.begin(),
+	           parser_result.found_lcd_fonts.end(),
+	           inserter(all_found, all_found.begin()));
+
+	std::string code_pages_str = {};
+	for (const auto item : all_found) {
+		if (!code_pages_str.empty()) {
+			code_pages_str += ", ";
+		}
+		code_pages_str += std::to_string(item);
+	}
+	LOG_WARNING("LOCALE: The file '%s' does not contain code page %d; it contains %s",
+	            file_name.c_str(),
+	            code_page,
+	            code_pages_str.c_str());
+
+	return KeyboardLayoutResult::NoCodePageInCpiFile;
+}
```

Comment:
Another huge method, a good candidate for splitting things up into more logical chunks.

Reply:
Extracted reading the file via MS-DOS API to external function; this largely reduced the function size.